[MHDP-10] added functionality to edit a person's affiliation

prnvshrn opened the pull request:
For this ticket I have added a new feature, registered users can go to <b>Affiliations</b> tab and edit the _position_ and _institution_ information.

Commit:
[MHDP-10] added functionality to edit a person's affiliation

jdamerow commented on browser/views.py:
```diff
@@ -686,3 +686,31 @@ def delete_investigator_record(request, person_id, research_id):
+                    institution = Institution.objects.get_or_create(
+                        name=form.cleaned_data.get('institution'),
+                        changed_by=request.user
+                    )
```
don't filter by name but by primary key

jdamerow replied:
names are not guaranteed to be unique. or someone else might change the name right before this is called. you would then create a new institution rather than reusing the existing one.

prnvshrn replied:
I have changed the edit_affiliation view, now id is being used to retrieve institution information.

Commit:
[MHDP-10] added check for creating new institutes

Commit:
[MHDP-10] Getting institution information by id rather than name

Commit:
[MHDP-10] Getting institution information by id rather than name

Commit:
[MHDP-10] feature for deleting affiliation

prnvshrn commented:
Made following changes -
1. id is being used to retrieve institution information
2. added feature of deleting affiliation

Commit:
[MHDP-10] Removed delete affiliation feature

prnvshrn commented:
I have removed the delete affiliation feature. This is because there are two models **Affiliation** and **Attendance** involved and both are not connected to each other. So, if I delete an **Affiliation** entry I will not be able to delete the corresponding **Attendance** entry and the deleted affiliation will show up in the Attendee section of a course.

The only information we can use from Affiliation is {year, person_id} to delete entry in Attendance. But I think that it would be incorrect.

prnvshrn commented:
Resolving merge conflicts

Commit:
[MHDP-10] Removed merge conflicts

Commit:
[MHDP-10] Fixed bug of adding blank institutes

jdamerow commented:
we just discussed this

Commit:
[MHDP-10] Corrected rendering issues of tabs on Person page

Commit:
[MHDP-10] added deleting affiliations

Commit:
[MHDP-10] corrected app filter

jdamerow commented on browser/views.py:
```diff
@@ -713,6 +713,60 @@ def delete_investigator_record(request, person_id, research_id):
+def delete_affiliation_record(request, person_id, affiliation_id):
+    affiliation = get_object_or_404(Affiliation, pk=affiliation_id)
+    if request.method == 'POST':
+        Affiliation.objects.select_related().filter(person_id=affiliation.person_id, year=affiliation.year
```
is there a reason you call 'select_related' here?

jdamerow commented on browser/views.py:
```diff
@@ -713,6 +713,60 @@ def delete_investigator_record(request, person_id, research_id):
+                else:
+                    institution = Institution.objects.get(id=form.cleaned_data.get('institution_id'))
+
+            Affiliation.objects.select_related().filter(person_id=affiliation.person_id, year=affiliation.year,
```
why do you call 'select_related' here?

jdamerow commented on browser/views.py:
```diff
@@ -713,6 +713,60 @@ def delete_investigator_record(request, person_id, research_id):
+                            changed_by=request.user
+                        )
+                else:
+                    institution = Institution.objects.get(id=form.cleaned_data.get('institution_id'))
```
this seems like the same code you had in your other PR. If so, this should not be duplicated but put in a helper method.

Commit:
[MHDP-10] checking for duplicates and added helper function

Commit:
[MHDP-10] added labels for multiple entries, created helper function for institutions

Commit:
[MHDP-10] Removed get-pip

jdamerow commented on browser/views.py:
```diff
@@ -713,6 +713,65 @@ def delete_investigator_record(request, person_id, research_id):
+        else:
+            messages.add_message(request, messages.ERROR,
+                                 'The above institute does not exist in the database. '
+                                 'Please check the checkbox below to create new institute')
```
"The above institution [...] to create a new institution."

jdamerow commented on browser/views.py:
```diff
@@ -713,6 +713,65 @@ def delete_investigator_record(request, person_id, research_id):
+                    )
+            else:
+                institution = Institution.objects.get(id=form.cleaned_data.get('institution_id'))
+        return {'status': True, 'institution': institution}
```
why is this called 'status' if its value is only False or True?

jdamerow replied:
just return a tuple and put the return value into two variables when calling the function

jdamerow commented on browser/views.py:
```diff
@@ -713,6 +713,65 @@ def delete_investigator_record(request, person_id, research_id):
+    return HttpResponseRedirect(reverse('person', args=(person_id,)))
+
+
+# This helper checks if an institution is to be created or an existing institution is to be retrieved
```
use the proper code style here

Commit:
[MHDP-10] addressed review comments:changed helper return method

jdamerow commented on browser/views.py:
```diff
@@ -713,6 +713,66 @@ def delete_investigator_record(request, person_id, research_id):
+    return HttpResponseRedirect(reverse('person', args=(person_id,)))
+
+
+def handle_institution(form, institution, request):
```
this is a private method and shoudl start with "_'

jdamerow replied:
why do you pass 'institution' into this method?

jdamerow commented on browser/views.py:
```diff
@@ -713,6 +713,66 @@ def delete_investigator_record(request, person_id, research_id):
+
+    if request.method == 'POST':
+        form = EditAffiliationForm(request.POST, instance=person)
+        check_handle_affiliation = handle_institution(form, None, request)
```
you should let python unpack the return value for you
```
a, b = function_that_returns_two_values()
```

Commit:
[MHDP-10] addressed review comments: edited handle_affiliation signature

jdamerow commented on browser/views.py:
```diff
@@ -713,6 +713,67 @@ def delete_investigator_record(request, person_id, research_id):
+        success, institution = _handle_institution(form, request)
+        if success:
+            Affiliation.objects.filter(id=affiliation_id).update(position=form.cleaned_data.get('position'),
+                                                     institution=institution)
```
is there a reason you don't use save() here? If there are any actions associated with saving an object (e.g. update last modified timestamp), these wont' be run if you do it this way.

jdamerow commented on browser/views.py:
```diff
@@ -713,6 +713,67 @@ def delete_investigator_record(request, person_id, research_id):
+    Checks if an institution is to be created or an existing institution is to be retrieved
+    """
+    institution = None
+    if form.is_valid():
```
what if the form is not valid?

prnvshrn replied:
If the form is not valid, it goes back to the parent class and displays the error on the form. I verified it by changing one of the form's required field to True and tried submitting it.

jdamerow replied:
ya, since you're not returning anything, success isn't true. But that's pretty bad practice as it is very unexpected.

Commit:
[MHDP-10] addressed review comments: changed update to save in editing affiliation

jdamerow commented on browser/views.py:
```diff
@@ -713,6 +713,69 @@ def delete_investigator_record(request, person_id, research_id):
+                # Checking if create institution radio button has been checked
+                if not form.cleaned_data.get('create_institution'):
+                    return False, institution
+                else:
```
no else needed, since if return.

jdamerow commented:
almost there

Commit:
[MHDP-10] addressed review comments: restructured _handle_institution

Commit:
[MHDP-10] added form in context

Commit:
[MHDP-10] set DEBUG to False

prnvshrn commented:
Made following changes :- 

- Removed else
- Moved form.is_valid to the parent function. Only if the form is valid, handle_institution will be called.
